In [1]:
from pathlib import Path

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
import sys
sys.executable

'c:\\temp\\python_learning\\ml_projects\\ml_projects_batch_01\\02_telco_customer_churn\\.venv\\Scripts\\python.exe'

In [3]:
PROJECT_ROOT = Path.cwd().parent
DATA_PATH = PROJECT_ROOT / "data" / "raw" / "WA_Fn-UseC_-Telco-Customer-Churn.csv"

PROJECT_ROOT, DATA_PATH

(WindowsPath('c:/temp/python_learning/ml_projects/ml_projects_batch_01/02_telco_customer_churn'),
 WindowsPath('c:/temp/python_learning/ml_projects/ml_projects_batch_01/02_telco_customer_churn/data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv'))

In [4]:
df = pd.read_csv(DATA_PATH)

df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [5]:
df.shape

(7043, 21)

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

In [7]:
df.columns.tolist()

['customerID',
 'gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'MonthlyCharges',
 'TotalCharges',
 'Churn']

In [8]:
df.isna().sum().sort_values(ascending=False)

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [9]:
df.duplicated().sum()

np.int64(0)

In [16]:
df["InternetService"].value_counts()

InternetService
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64

In [11]:
df["Churn"].value_counts(normalize=True)

Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64

Предварительно target: churn. Бинарный таргет.
Предварительно ID column: customerID, содержит ID пользователей и не несет смысла для моделей.
Подозрительные колонки по dtype: почти со всеми колонками придется работать - переводить в нужный формат. Например, SeniorCitizen уже в бинарном формате (0,1), MonthlyCharges и TotalCharges в формате str, хотя по факту являются float, многоклассовая классификация точно у MultipleLines, InternetService, PaymentMethod.

In [17]:
df.nunique().sort_values(ascending=False)

customerID          7043
TotalCharges        6531
MonthlyCharges      1585
tenure                73
PaymentMethod          4
StreamingMovies        3
TechSupport            3
OnlineBackup           3
StreamingTV            3
DeviceProtection       3
MultipleLines          3
InternetService        3
OnlineSecurity         3
Contract               3
Partner                2
SeniorCitizen          2
gender                 2
Dependents             2
PhoneService           2
PaperlessBilling       2
Churn                  2
dtype: int64

In [19]:
pd.set_option('display.max_rows', None)
for col in df.columns:
    print(f"\n{col}")
    print(df[col].unique()[:20])


customerID
<StringArray>
['7590-VHVEG', '5575-GNVDE', '3668-QPYBK', '7795-CFOCW', '9237-HQITU',
 '9305-CDSKC', '1452-KIOVK', '6713-OKOMC', '7892-POOKP', '6388-TABGU',
 '9763-GRSKD', '7469-LKBCI', '8091-TTVAX', '0280-XJGEX', '5129-JLPIS',
 '3655-SNQYZ', '8191-XWSZG', '9959-WOFKT', '4190-MFLUW', '4183-MYFRB']
Length: 20, dtype: str

gender
<StringArray>
['Female', 'Male']
Length: 2, dtype: str

SeniorCitizen
[0 1]

Partner
<StringArray>
['Yes', 'No']
Length: 2, dtype: str

Dependents
<StringArray>
['No', 'Yes']
Length: 2, dtype: str

tenure
[ 1 34  2 45  8 22 10 28 62 13 16 58 49 25 69 52 71 21 12 30]

PhoneService
<StringArray>
['No', 'Yes']
Length: 2, dtype: str

MultipleLines
<StringArray>
['No phone service', 'No', 'Yes']
Length: 3, dtype: str

InternetService
<StringArray>
['DSL', 'Fiber optic', 'No']
Length: 3, dtype: str

OnlineSecurity
<StringArray>
['No', 'Yes', 'No internet service']
Length: 3, dtype: str

OnlineBackup
<StringArray>
['Yes', 'No', 'No internet service']
Length:

In [23]:
string_cols = df.select_dtypes(include=["object", "string"]).columns.tolist()

for col in string_cols:
    n_blank = (df[col].str.strip() == "").sum()
    if n_blank > 0:
        print(col, n_blank)

TotalCharges 11


In [24]:
df.loc[df["TotalCharges"].str.strip() == "", ["customerID", "tenure", "MonthlyCharges", "TotalCharges", "Churn"]]

,customerID,tenure,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,0,52.55,,No
753,3115-CZMZD,0,20.25,,No
936,5709-LVOEQ,0,80.85,,No
1082,4367-NUYAO,0,25.75,,No
1340,1371-DWPAZ,0,56.05,,No
3331,7644-OMVMY,0,19.85,,No
3826,3213-VVOLG,0,25.35,,No
4380,2520-SGTTA,0,20.00,,No
5218,2923-ARZLG,0,19.70,,No
6670,4075-WKNIU,0,73.35,,No


In [25]:
total_charges_numeric = pd.to_numeric(df["TotalCharges"], errors="coerce")

total_charges_numeric.isna().sum()

np.int64(11)

In [27]:
df.loc[total_charges_numeric.isna(), ["customerID", "tenure", "MonthlyCharges", "TotalCharges", "Churn"]]

,customerID,tenure,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,0,52.55,,No
753,3115-CZMZD,0,20.25,,No
936,5709-LVOEQ,0,80.85,,No
1082,4367-NUYAO,0,25.75,,No
1340,1371-DWPAZ,0,56.05,,No
3331,7644-OMVMY,0,19.85,,No
3826,3213-VVOLG,0,25.35,,No
4380,2520-SGTTA,0,20.00,,No
5218,2923-ARZLG,0,19.70,,No
6670,4075-WKNIU,0,73.35,,No


TotalCharges should be treated as numeric feature.
It has 11 hidden missing values caused by blank strings.
These rows correspond to customers with tenure = 0.

In [28]:
target_col = "Churn"

id_cols = ["customerID"]

numeric_cols_initial = [
    "SeniorCitizen",
    "tenure",
    "MonthlyCharges",
    "TotalCharges",
]

categorical_cols_initial = [
    col for col in df.columns
    if col not in numeric_cols_initial + id_cols + [target_col]
]

target_col, id_cols, numeric_cols_initial, categorical_cols_initial

('Churn',
 ['customerID'],
 ['SeniorCitizen', 'tenure', 'MonthlyCharges', 'TotalCharges'],
 ['gender',
  'Partner',
  'Dependents',
  'PhoneService',
  'MultipleLines',
  'InternetService',
  'OnlineSecurity',
  'OnlineBackup',
  'DeviceProtection',
  'TechSupport',
  'StreamingTV',
  'StreamingMovies',
  'Contract',
  'PaperlessBilling',
  'PaymentMethod'])

In [29]:
len(numeric_cols_initial), len(categorical_cols_initial)

(4, 15)

In [30]:
for col in categorical_cols_initial:
    print(f"\n{col}: {df[col].nunique()} unique values")
    print(df[col].value_counts(dropna=False))


gender: 2 unique values
gender
Male      3555
Female    3488
Name: count, dtype: int64

Partner: 2 unique values
Partner
No     3641
Yes    3402
Name: count, dtype: int64

Dependents: 2 unique values
Dependents
No     4933
Yes    2110
Name: count, dtype: int64

PhoneService: 2 unique values
PhoneService
Yes    6361
No      682
Name: count, dtype: int64

MultipleLines: 3 unique values
MultipleLines
No                  3390
Yes                 2971
No phone service     682
Name: count, dtype: int64

InternetService: 3 unique values
InternetService
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64

OnlineSecurity: 3 unique values
OnlineSecurity
No                     3498
Yes                    2019
No internet service    1526
Name: count, dtype: int64

OnlineBackup: 3 unique values
OnlineBackup
No                     3088
Yes                    2429
No internet service    1526
Name: count, dtype: int64

DeviceProtection: 3 unique values
DeviceProtecti

In [31]:
audit_summary = {
    "n_rows": df.shape[0],
    "n_columns": df.shape[1],
    "target": target_col,
    "id_columns": id_cols,
    "numeric_features_initial": numeric_cols_initial,
    "categorical_features_initial": categorical_cols_initial,
    "hidden_missing_total_charges": int((df["TotalCharges"].str.strip() == "").sum()),
    "duplicates": int(df.duplicated().sum()),
}

audit_summary

{'n_rows': 7043,
 'n_columns': 21,
 'target': 'Churn',
 'id_columns': ['customerID'],
 'numeric_features_initial': ['SeniorCitizen',
  'tenure',
  'MonthlyCharges',
  'TotalCharges'],
 'categorical_features_initial': ['gender',
  'Partner',
  'Dependents',
  'PhoneService',
  'MultipleLines',
  'InternetService',
  'OnlineSecurity',
  'OnlineBackup',
  'DeviceProtection',
  'TechSupport',
  'StreamingTV',
  'StreamingMovies',
  'Contract',
  'PaperlessBilling',
  'PaymentMethod'],
 'hidden_missing_total_charges': 11,
 'duplicates': 0}

## Stage 0 findings

- Dataset shape: 7043 rows, 21 columns.
- Target: `Churn`.
- Task type: binary classification.
- Positive class: `Churn = Yes`.
- Class distribution:
  - No: 5174 / 73.46%
  - Yes: 1869 / 26.54%
- ID column: `customerID`.
- Explicit missing values: 0.
- Full duplicate rows: 0.
- Hidden missing values:
  - `TotalCharges` has 11 blank values.
- `TotalCharges` is semantically numeric but loaded as string.
- Initial numeric features:
  - `SeniorCitizen`
  - `tenure`
  - `MonthlyCharges`
  - `TotalCharges`
- Initial categorical features:
  - all non-ID, non-target, non-numeric service/customer profile columns.
- Initial leakage-risk assessment:
  - no obvious direct leakage columns found at this audit stage.
  - `customerID` must be excluded because it is an identifier.

## Task framing

We solve a binary classification problem.

Goal: predict whether a customer is likely to churn.

Target column: `Churn`.

Target values:
- `No`: customer did not churn.
- `Yes`: customer churned.

Positive class: `Yes`.

Business interpretation:
- False Positive: model predicts churn, but the customer would not actually churn.
  - Possible cost: unnecessary retention offer, discount, or sales effort.
- False Negative: model predicts no churn, but the customer actually churns.
  - Possible cost: lost customer and lost future revenue.

For churn prediction, recall for the positive class is important because the business usually wants to detect as many at-risk customers as possible.

Precision is also important because too many false positives can waste retention budget.

Accuracy alone is not enough because the dataset is imbalanced:
- majority class `No` is about 73.46%;
- a naive model predicting only `No` would already get about 73.46% accuracy.

Initial main metrics:
- Recall for `Churn = Yes`
- Precision for `Churn = Yes`
- F1-score for `Churn = Yes`
- ROC-AUC as a threshold-independent ranking metric

Later we may also look at:
- confusion matrix
- PR-AUC
- threshold tuning on validation data, not on test data

In [32]:
majority_class_accuracy = df["Churn"].value_counts(normalize=True).max()
majority_class_accuracy

np.float64(0.7346301292063041)

In [33]:
positive_rate = (df["Churn"] == "Yes").mean()
positive_rate

np.float64(0.2653698707936959)